src/mongoose_api_json.erl

```diff
@@ -38,39 +38,49 @@ serialize(Data) ->
 %%--------------------------------------------------------------------
 %% internal functions
 %%--------------------------------------------------------------------
-do_deserialize({ElementName, {struct, [{_Key, _Value}|_Rest]=Proplist}}) ->
-    {ElementName, do_deserialize(Proplist)};
-do_deserialize({struct, Proplist}) ->
-    [do_deserialize(Element) || Element <- Proplist];
-do_deserialize(List) when is_list(List) ->
-    [do_deserialize(Element) || Element <- List];
-do_deserialize(Other) ->
-    Other.
+do_deserialize(#{} = Map) ->
+    maps:to_list(maps:map(fun(_K, V) -> do_deserialize(V) end, Map));
+do_deserialize(NotAMap) ->
+    NotAMap.
 
 do_serialize(Data) ->
-    mochijson2:encode(prepare_struct(Data)).
+    jiffy:encode(prepare_struct(Data)).
 
 prepare_struct({Key, Value}) ->
-    {struct, [{Key, prepare_struct(Value)}]};
+    #{prepare_key(Key) => prepare_struct(Value)};
 prepare_struct([]) ->
     [];
 prepare_struct(List) when is_list(List) ->
     case is_proplist(List) of
         true  ->
-            Fields = [{K, prepare_struct(V)} || {K, V} <- List],
-            {struct, Fields};
+            maps:from_list([{prepare_key(K), prepare_struct(V)} || {K, V} <- List]);
         false ->
             [prepare_struct(Element) || Element <- List]
     end;
 prepare_struct(Other) ->
+    prepare_str(Other).
+
+prepare_key(Key) when is_integer(Key) ->
+    integer_to_binary(Key);
+prepare_key(Key) ->
+    prepare_str(Key).
+
+prepare_str(List) when is_list(List) ->
+    try unicode:characters_to_binary(List) of
+        Bin when is_binary(Bin) -> Bin;
+        _ -> List
+    catch
+        error:badarg -> List
```

Comment:
Should we fall back to original list here? Maybe we should crash if we don't get a valid unicode here?

Reply:
This is a kludge to try and encode lists that are actually strings - lists that are actually lists should be encoded as JSON arrays instead

Reply:
To rephrase, we don't know if the list is a string or not - so we try to encode it to check.

Reply:
Please rename this function then or add a comment because this semantics is not obvious.